fix(datepicker): remove date-fns and allow more customization

redallen opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Remove `date-fns` and amend API to be more flexible given feedback from @nicolethoen and @karelhala .

- Change default formatting/parsing to `yyyy-MM-dd`.
  - Allows relying on default browser formatting/parsing in DatePicker via native date methods like `date.toISOString().substring(0, 10)` and ``new Date(`${val}T00:00:00`)``.
  - Allows CalendarMonth to render any ReactNodes via format functions:
     ```js
     monthFormat = date => date.toLocaleDateString(locale, { month: 'long' }),
     weekdayFormat = date => date.toLocaleDateString(locale, { weekday: 'narrow' }),
     longWeekdayFormat = date => date.toLocaleDateString(locale, { weekday: 'long' }),
     buttonFormat = date => date.getDate()
     ```
      - `locale` defaults to the browser's locale giving decent i18n out of the box. We can force `en-US` to make our Calendar look consistent regardless of language.
- Expose new `weekStart` prop
- Remove `minDate`, `maxDate`, `invalidFormatErrorMessage`, `dateOutOfRangeErrorMessage`, `beforeMinDateErrorMessage`, and `afterEndDateErrorMessage` in favor of `validators?: ((date: Date) => string)[];`. This is more elegant and allows greater flexibility moving forward.

Closes #5142
Closes #5156

<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-5161.surge.sh

nicolethoen reviewed (commented):
exposing third party libraries is something we'd normally avoid, right?

redallen commented:
Yes and the reason for that is we don't want consumers depending too heavily on 3rd party deps that we might modify or swap in the future.

In this particular case unless we pick a date string formatting spec and abstract `date-fns` away with our own `format` and `parse` methods I don't think we can modify or swap `date-fns` in the future. I don't foresee a reason to swap or modify it since they support a wide range of strings and locales.

FWIW, we're already "exposing" the Locales from `date-fns`. If we really wanted to not expose this third party library we'd have to allow the consumer to define their own `format` and `parse` functions which also handle locales for both CalendarMonth and DatePicker. I'm not opposed to that given that many older PatternFly consumers use Moment, but it is an annoyance that'd require 2 new props and to hook up an existing or new library.

karelhala reviewed (approved):
Hella sweet! I like how you cleaned the `DatePicker`, I know it brings the heavy lifting to consumer, but this feels more robust and straight forward. I'd just change the comments above format props from `{propName} according to date-fns` to for instance `Format month from given Date, used in Select`

jschuler commented on packages/react-datetime/src/components/CalendarMonth/CalendarMonth.tsx:
```diff
@@ -6,42 +6,41 @@ import ArrowLeftIcon from '@patternfly/react-icons/dist/js/icons/arrow-left-icon
+  /** If using the default formatters which locale to use. Undefined defaults to current locale. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#Locale_identification_and_negotiation */
+  locale?: string;
+  /** Day of week that starts the week. 0 is Sunday, 6 is Saturday. */
+  weekStart?: 0 | 1 | 2 | 3 | 4 | 5 | 6;
```
Might be nicer to see `SUN | MON | TUE | WED | THU | FRI | SAT` for less mental mapping

redallen replied:
I've added a `Weekday` enum.

jschuler commented on packages/react-datetime/src/components/DatePicker/DatePicker.tsx:
```diff
@@ -3,165 +3,108 @@ import { css } from '@patternfly/react-styles';
+  /** How to format days in Calendar */
+  longWeekdayFormat?: (date: Date) => React.ReactNode;
+  /** How to format days in Calendar */
+  buttonFormat?: (date: Date) => React.ReactNode;
```
Same here, `dayFormat`?

jschuler reviewed (commented):
CalendarMonth and DatePicker share many props, I think they can extend out a common interface

nicolethoen commented on packages/react-datetime/src/components/DatePicker/DatePicker.tsx:
```diff
@@ -3,165 +3,108 @@ import { css } from '@patternfly/react-styles';
-  placeholder = 'MM/dd/yyyy',
-  value: valueProp = '',
+  placeholder = 'yyyy-MM-dd',
+  value: valueProp = dateFormat(new Date()),
```
I think it should be empty rather than have it default to the current date if no default date is provided.

redallen replied:
Done.

nicolethoen reviewed (commented):
How hard would it be for someone to configure it so that the date is more readable like most users are accustomed? I think we should be sure that it's very easy, since it will be very common for people to expect the date format they see most often.

redallen commented:
@jschuler I made a common "CalendarFormat" interface that appears in the docs, renamed `buttonFormat` -> `dayFormat`, and added a `Weekday` enum.

@nicolethoen I added an "American" example where I specify a simple `dateFormat` and `dateParse` using native APIs that works for 4-digit years. I made the default value an empty string.

nicolethoen commented:
nit: can you make maybe the basic date picker example - or at least one of the examples - not have a default date? :)
